cadc-storage-adapter-fs/src/main/java/org/opencadc/inventory/storage/fs/FileSystemStorageAdapter.java

```diff
@@ -0,0 +1,523 @@
+            wrapper.write(digestOut);
+            digestOut.flush();
+            byte[] md5sum = digest.digest();
+            String md5Val = HexUtil.toHex(md5sum);
+            checksum = URI.create(MD5_CHECKSUM_SCHEME + ":" + md5Val);
+            length = Files.size(path);
+            log.debug("calculated md5sum: " + checksum);
+            log.debug("calculated file size: " + length);
+        } catch (Throwable t) {
+            throwable = t;
+            log.error("error writing file", t);
+        } finally {
+            if (existing != null) {
+                if (throwable != null) {
+                    try {
+                        log.debug("Restoring original file.");
+                        Files.move(existing, path, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
+                    } catch (IOException e) {
+                        log.error("Failed to restore original file", e);
+                    }
+                } else {
+                    try {
+                        log.debug("Deleting original file.");
+                        Files.delete(existing);
+                    } catch (IOException e) {
+                        log.error("Failed to delete original file", e);
+                    }
+                }
+            }
+            
+            if (throwable != null) {
+                if (throwable instanceof IOException) {
+                    throw (IOException) throwable;
+                }
+                // TODO: identify throwables that are transient
+                throw new IllegalStateException("Unexpected error", throwable);
+            }
+        }
+        
+        // checksum comparison
+        if (newArtifact.contentChecksum != null && newArtifact.contentChecksum.getScheme().equals(MD5_CHECKSUM_SCHEME)) {
+            String expectedMD5 = newArtifact.contentChecksum.getSchemeSpecificPart();
+            String actualMD5 = checksum.getSchemeSpecificPart();
+            if (!expectedMD5.equals(actualMD5)) {
+                throw new StreamCorruptedException(
+                    "expected md5 checksum [" + expectedMD5 + "] "
+                    + "but calculated [" + actualMD5 + "]");
+            }
+        } else {
+            log.debug("Uncomparable or no contentChecksum provided.");
+        }
+        
+        // content length comparison
+        if (newArtifact.contentLength != null) {
+            Long expectedLength = newArtifact.contentLength;
+            if (!expectedLength.equals(length)) {
+                throw new StreamCorruptedException(
+                    "expected contentLength [" + expectedLength + "] "
+                    + "but calculated [" + length + "]");
+            }
+        } else {
+            log.debug("No contentLength provided.");
+        }
+        
+        URI storageID = null;
+        String storageBucket = null;
+        switch (bucketMode) {
+            case URI_BASED:
```

Comment:
Part of the URI_BASED case is the same sequence of appends as in createArtifactPath. That might be easier to not duplicate if createArtifactPath was createStorageLocation and returned a StorageLocation. Then the same code as used bu get() can be used to actually get the Path to write to.

Reply:
Took me a while to understand this suggestion but got it now.  I've refactored put so it creates the storage location upfront and uses the existing method to get the path from the location.  It fell out nicely.